### PR TITLE
Hide start position and faction when they are random

### PR DESCRIPTION
**Pull request by Crotalus**
Rebased #1215 

TL;DR
- Shuffle positions within a team (sim-side) to prevent UI mods of showing start position of random spawns
- Hide random faction until a player is scouted

There are UI mods floating around which shows start position of a player by looking at the scenario files for ArmyID ->  position. Another easy way is to press F11 if you memorized the positions (Setons - 1, 2 for front .. 7, 8 air)

<Link href="http://forums.faforever.com/viewtopic.php?f=41&t=11577" target="_blank">http://forums.faforever.com/viewtopic.php?f=41&t=11577</Link>

With random / balanced spawns this should be considered an exploit. By shuffling the team positions sim side we prevent user side to pin down the start positions.

Another improvement could be a game option "Incognito" which completely hide the names of the enemies (show in replay / if players choose to reveal themselves during game). Random colors then also ofc. Currently high ranked players are ganked (and vice versa with weak players). Would be better if such strategies were a result of scouting an opponent's gameplay, not simply by looking at the commander name.

### CODE REVIEW

File: lua/aibrain.lua [hunk 1]
```diff
@@ -47,6 +47,9 @@ local Points = {
 AIBrain = Class(moho.aibrain_methods) {
     Result = nil,
 
+    -- unit:GetArmy() so why brain:GetArmyIndex() ?!
+    GetArmy = moho.aibrain_methods.GetArmyIndex,
```
Comment (aeoncleanse) [inline lua/aibrain.lua]: What???

Comment (Sheeo) [inline lua/aibrain.lua]: Not a very helpful question @aeoncleanse. What do you mean, what?

Comment (aeoncleanse) [inline lua/aibrain.lua]: Sorry got distracted shortly after typing this. I just don't understand at all what this is doing, or why.

Comment (Crotalus) [inline lua/aibrain.lua]: It's a function reference so you can do brain:GetArmy() in the same way as you do unit:GetArmy(), just for convenience

Comment (aeoncleanse) [inline lua/aibrain.lua]: Ahhhhhh. Ok thanks, that makes sense now

**Comment by aeoncleanse** [comment]
@Crotalus I need step by step instructions on how to test this. @duk3luk3 @Uveso you guys are both better programmers than I am. I'd like both of you to go over this please, it's a complex system.

**Comment by Crotalus** [comment]
1. Choose a map where you know the numbering of start positions. For setons it is (1 Front, 2 Front, 3 Rock, 4 Rock, 5 Beach, 6 Beach, 7 Back, 8 Back).
2. Choose random (or autobalance) start position in lobby, preferabbly with an enabled "reveal location UI" mod which reveal the player name of each position.
3. Choose random faction for at least one enemy player
4. When in game, press F11 to see the list of players, see if the start position no longer correlates to these. The UI mod which writes out names should now be reporting wrong positions (unless it was lucky).
5. Check that non-scouted enemies who are random faction are shown as such in the score tab.
6. When you scout the an enemy player, make sure the random faction icon change into the actual faction of that player.
7. Rince and repeat, try with AI players

Find other bugs and try to break as much as possible.

**Comment by Uveso** [comment]
Is it possible to move the `Crotalus:hide-startpos` branch to `FAForever/fa:hide-startpos` ?
This way its more easy to test this for some people like me :)

**Comment by Crotalus** [comment]
```
git checkout -b Crotalus-hide-startpos develop
git pull https://github.com/Crotalus/fa.git hide-startpos
```

**Comment by Uveso** [comment]
@Crotalus 
Thanks a lot for the 2 command lines.
Worked perfect!

***

I started testing, and ended in skirmish screen.
if i press start, i get an error message:

```
WARNING: Error running OnRelease script in <deleted object>: ...programdata\faforever\repo\fa\lua\ui\lobby\lobby.lua(1482): bad argument #1 to `getn' (table expected, got nil)
         stack traceback:
         	[C]: in function `getn'
         	...programdata\faforever\repo\fa\lua\ui\lobby\lobby.lua(1482): in function <...programdata\faforever\repo\fa\lua\ui\lobby\lobby.lua:1470>
         	...programdata\faforever\repo\fa\lua\ui\lobby\lobby.lua(1699): in function `LaunchGame'
         	...programdata\faforever\repo\fa\lua\ui\lobby\lobby.lua(1731): in function <...programdata\faforever\repo\fa\lua\ui\lobby\lobby.lua:1588>
         	...programdata\faforever\repo\fa\lua\ui\lobby\lobby.lua(2836): in function `OnClick'
         	c:\programdata\faforever\repo\fa\lua\maui\button.lua(85): in function <c:\programdata\faforever\repo\fa\lua\maui\button.lua:78>

```

To reproduce the error;
Use one AI with random faction. and you can't start.
Also Multiplayer, the same crash in line 1482 if you try to press the start button

**Comment by Crotalus** [comment]
Thanks for the reports, I've updated the code to support random faction AI better and rebased to latest `develop`.

**Comment by Uveso** [comment]
If you play with an AI with Omni-settings on, you can see all factions from start on.

If you disable the omni for the AI, the AI will not start to build something.

**Comment by Crotalus** [comment]
Fixed bug with AI not building from start and rebased.

As to the AI omni cheat, if you have such an AI on your team it will see / scout the enemy thus knowing the faction. I guess we could make an exception that all-seeing AI should be excluded.

**Comment by aeoncleanse** [comment]
Likewise @Crotalus please rebase this onto deploy/fafdevelop

**Comment by aeoncleanse** [comment]
Given that the server is notified of Army Index on game launch, where the index is done in slot order (Missing empty slots), will this PR mess with that? As in, will this result in army indexes being shuffled post-game launch?

**Comment by aeoncleanse** [comment]
Refreshed Travis

**Comment by aeoncleanse** [comment]
Rebased new PR #1810
